File: binarytree.py
class BinaryTree:

    # 5. constructor to initialize binary tree as an empty tree or with a single value
    def __init__(self, data=None, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right

    # 6. add_leftchild and add_rightchild methods
    # 7. TypeError if new child is not of the same type as the rest of the elements of the tree
    def add_leftchild(self, left):
        if isinstance(left.data, type(self.data)) or left.data is None:
            self.left = left
        else:
            raise TypeError("Type mismatch between " + str(type(self.data)) + " and " + str(type(left)))

    def add_rightchild(self, right):
        if isinstance(right.data, type(self.data)) or right.data is None:
            self.right = right
        else:
            raise TypeError("Type mismatch between " + str(type(self.data)) + " and " + str(type(right)))

    # 8. property and setter
    @property
    def data(self):
        return self.__data

    @data.setter
    def data(self, data):
        self.__data = data

    # 9. implementation of iter()
    def __iter__(self):
        if self.data is not None or self.data is None and isinstance(self, BinaryTree):
            yield self.data
        if self.left is not None:
            for d in self.left:
                yield d
        if self.right is not None:
            for d in self.right:
                yield d

File: BinaryTree.java
import java.util.*;

public class BinaryTree<T> implements Iterable<T>{

    public Node<T> root;

    //#1 node-and-pointer implementation of binary tree
    static class Node<T> {
        T data;
        BinaryTree<T> right = null, left = null;

        Node(T data) {this.data = data;}
    }

    //#2 no-arg constructor for empty tree, and constructor for tree with single node containing a specific value
    public BinaryTree() {};

    public BinaryTree(T data) {setRoot(data);}

    public BinaryTree(BinaryTree<T> left, T data, BinaryTree<T> right) {
        setRoot(data);
        addLeftChild(left);
        addRightChild(right);
    }

    public void setRoot(T data) {root = new Node<>(data);}

    //#3 addLeftChild and addRightChild
    public void addLeftChild(BinaryTree<T> tree) {root.left = tree;}

    public void addRightChild(BinaryTree<T> tree) {root.right = tree;}

    @Override
    public Iterator<T> iterator() {
        return new BinaryTreeIterator<>(this);
    }

    //#4 iterator implementation
    private static class BinaryTreeIterator<T> implements Iterator<T> {
       Queue<BinaryTree<T>> q = new LinkedList<>();

        BinaryTreeIterator(BinaryTree<T> tree) {
            if (tree.root.data != null)
                q.add(tree);
        }

        @Override
        public boolean hasNext() {return !q.isEmpty();}

        @Override
        public T next() {
            BinaryTree<T> tree = q.remove();
            //Node<T> node = tree.root;
            if (tree.root.left != null)
                q.add(tree.root.left);
            if (tree.root.right != null)
                q.add(tree.root.right);
            return tree.root.data;
        }
    }
}
